README.md
# agario-client Bot

## Instructions ##
- Install [Node.js](https://nodejs.org/)
- Install *"npm install ws"* from the command line
- Run `agario-devtools/repeater.js`
- Open [agar.io](http://agar.io) in Google Chrome
- Open de console by pressing F12
- Input the instruction given by repeater.js
- now start `agario-client/RoboBot.js`
- Profit

agario-client/CHANGELOG.md
## 08.06.2015 ##
Protocol changes:

- New packet id 240 that moves offset (why, agar? what for?)

Code changes:

- New packet management architecture
- [buffer-dataview](https://github.com/TooTallNate/node-buffer-dataview) not used anymore

## 07.06.2015 ##
`agario-client` added to [NPM](https://www.npmjs.com/package/agario-client)

## 06.06.2015 ##
Code changes:

- `Client.score` added (by [GeoffreyFrogeye](https://github.com/GeoffreyFrogeye))
- `Client.on.scoreUpdate(old_score, new_score)` added (by [GeoffreyFrogeye](https://github.com/GeoffreyFrogeye))

## 04.06.2015 ##
Code changes:

- `Ball.color` is now working (fixed by [GeoffreyFrogeye](https://github.com/GeoffreyFrogeye))
- New events methods (improved by [GeoffreyFrogeye](https://github.com/GeoffreyFrogeye))
- Deprecated property `Ball.is_virus` completely removed
- Deprecated property `Ball.is_mine` completely removed
- `.off()` marked as deprecated and replaced with `.removeListener()`
- `.offAll()` marked as deprecated and replaced with `.removeAllListeners()`
- `.emitEvent()` marked as deprecated and replaced with `.emit()`
- `Client.server` added

## 01.06.2015 ##
Protocol changes:

- `ball` coordinates changed from 32bit float to 16bit signed integer
- `ball` size changed from 32bit float to 16bit signed integer
- packet ID 16 part 3 changed from list of visible balls to list of destroyed balls
- two bits between 2 and 3 part of packet 16 is not sent anymore

## 18.05.2015 ##
Now `example.js` will automatically request server and connect to it.

## 12.05.2015 ##
Protocol changes:

- `ball` coordinates changed from 64bit float to 32bit float
- `ball` size changed from 64bit float to 32bit float
- color is now generating on server and sent to client
- new packet 72 that not used in original code
- new packet 50 used for teams scores in teams mode

Code changes:

- color is now stored in `Ball.color`
- added empty processor for packet ID 72 (packet not used in original code)
- added `Client.teams_scores` property for teams mode
- added `Client.on.teamsScoresUpdate(old_scores, new_scores)` event for teams mode

graphics/canvas.js
// ==UserScript==
// @name        RoboBot
// @namespace   RoboBot
// @include     http://agar.io/
// @version     3.04
// @require http://code.jquery.com/jquery-latest.js
// @grant       none
// @author      Kris Minkjan
// ==/UserScript==

$(function(){
    connect("ws://127.0.0.1:415/");
    var canvas = $("#canvas");
    var ctx = canvas[0].getContext('2d');

    var socket= new WebSocket('ws://127.0.0.1:8000/');
    socket.onopen= function() {
        socket.send('hello');
    };
    socket.onmessage= function(s) {
        console.log(s);
        //var a = JSON.parse(s.data);
        //console.log(a[0] + " : "+ a[1]);
        //ctx.strokeStyle = "#df4b26";
        //ctx.fillStyle = "#FF0000";
        //ctx.lineJoin = "round";
        //ctx.lineWidth = 5;
        //ctx.fillRect(a[0],a[1],150,75);
        //ctx.fillRect(10,10,150,75);
    };
});

agario-client/packet.js
//this file is for internal use with packets

function Packet(e) {
    if(e instanceof Buffer) {
        this.data = e;
        this.length = this.data.length;
    }else if((typeof Buffer) != 'undefined' && e.data instanceof Buffer) {
        this.data = e.data;
        this.length = this.data.length;
    }else{
        this.data = new DataView(e.data);
        this.length = this.data.byteLength;
    }

    this.offset = 0;
}

Packet.prototype = {
    readUInt8: function(p) {
        var offset = p || this.offset;
        var ret;
        if(this.data.getUint8) {
            ret = this.data.getUint8(offset);
        }else{
            ret = this.data.readUInt8(offset);
        }
        if(p === undefined) this.offset += 1;

        return ret;
    },

    readUInt16BE: function(p) {
        var offset = p || this.offset;
        var ret;
        if(this.data.getUint16) {
            ret = this.data.getUint16(offset, false);
        }else{
            ret = this.data.readUInt16BE(offset);
        }
        if(p === undefined) this.offset += 2;

        return ret;
    },

    readUInt16LE: function(p) {
        var offset = p || this.offset;
        var ret;
        if(this.data.getUint16) {
            ret = this.data.getUint16(offset, true);
        }else{
            ret = this.data.readUInt16LE(offset);
        }
        if(p === undefined) this.offset += 2;

        return ret;
    },

    readSInt16LE: function(p) {
        var offset = p || this.offset;
        var ret;
        if(this.data.getInt16) {
            ret = this.data.getInt16(offset, true);
        }else{
            ret = this.data.readInt16LE(offset);
        }
        if(p === undefined) this.offset += 2;

        return ret;
    },

    readUInt32LE: function(p) {
        var offset = p || this.offset;
        var ret;
        if(this.data.getUint32) {
            ret = this.data.getUint32(offset, true);
        }else{
            ret = this.data.readUInt32LE(offset);
        }
        if(p === undefined) this.offset += 4;

        return ret;
    },

    readUInt32BE: function(p) {
        var offset = p || this.offset;
        var ret;
        if(this.data.getUint32) {
            ret = this.data.getUint32(offset, false);
        }else{
            ret = this.data.readUInt32BE(offset);
        }
        if(p === undefined) this.offset += 4;

        return ret;
    },

    readFloat32LE: function(p) {
        var offset = p || this.offset;
        var ret;
        if(this.data.getFloat32) {
            ret = this.data.getFloat32(offset, true);
        }else{
            ret = this.data.readFloatLE(offset);
        }
        if(p === undefined) this.offset += 4;

        return ret;
    },

    readFloat32BE: function(p) {
        var offset = p || this.offset;
        var ret;
        if(this.data.getFloat32) {
            ret = this.data.getFloat32(offset, false);
        }else{
            ret = this.data.readFloatBE(offset);
        }
        if(p === undefined) this.offset += 4;

        return ret;
    },

    readFloat64LE: function(p) {
        var offset = p || this.offset;
        var ret;
        if(this.data.getFloat64) {
            ret = this.data.getFloat64(offset, true);
        }else{
            ret = this.data.readDoubleLE(offset);
        }
        if(p === undefined) this.offset += 8;

        return ret;
    },

    readFloat64BE: function(p) {
        var offset = p || this.offset;
        var ret;
        if(this.data.getFloat64) {
            ret = this.data.getFloat64(offset, false);
        }else{
            ret = this.data.readDoubleBE(offset);
        }
        if(p === undefined) this.offset += 8;

        return ret;
    },

    toString: function() {
        var out = '';
        for(var i=0;i<this.length;i++) {
            if(out) out += ' ';
            var char = this.readUInt8(i).toString(16);
            if(char.length == 1) out += '0';
            out += char;
        }

        return out;
    }
};

module.exports = Packet;

agario-client/RoboBot.js
/**
 * Created by Kris on 12-6-2015.
 */

var region = 'EU-London'; //server region to request

var http = require('http');
var AgarioClient = require('./agario-client.js'); //in your code you should do require('agario-client')
//var Repeater = require('./repeater.js'); //in your code you should do require('agario-client')

var client = new AgarioClient('worker'); //create new client and call it "worker" (not nickname)
//var repeater = new  Repeater();

var interval_id = 0; //here we will store setInterval's ID
var repeaterOn = false;

var canvas = null;
var WebSocketServer = require('ws').Server
    , wss = new WebSocketServer({port: 8000});
wss.on('connection', function(ws) {
    canvas = ws;
    console.log("connected, ");
    ws.on('message', function(message) {
        console.log('received: %s', message);
    });
    ws.send("0,0");
    start();
});

client.connect("ws://127.0.0.1:9158/");

client.on('connected', function () { //when we connected to server
    start();
});

function start() {
    client.log('spawning');
    client.spawn('[xFake]'); //spawning new ball
    interval_id = setInterval(update, 100);
}

function update() { // TODO check amount of times i've split
    // TODO dont run in to another cell when fleeing for a big on
    // TODO or just ignore really big ones for now
    //console.time('update');
    var my_ball = client.balls[client.my_balls[0]];
    if (!my_ball) return;

    /* Sort balls */
    var threats = [], targets = [], virusses = [], foods = [];
    for (var ball_id in client.balls) { //we go true all balls we know about
        var ball = client.balls[ball_id];
        if (!ball.visible || ball.mine) continue;

        if (ball.virus) {
            virusses.push(ball);
            continue;
        }
        if (ball.name == null && ball.size < 30) { // Safety
            foods.push(ball);
            continue;
        }
        var relativeSize = ball.size / my_ball.size;
        if (relativeSize > 1.1) { // I can be eaten
            threats.push(ball);
        } else if (relativeSize < 0.7) {
            targets.push(ball);
        }
    }

    /* Save all coordinates to run to */
    var run_x = [];
    var run_y = [];

    /* Save the unsafe directions */
    var bad_directions = [];

    /* Process threats */
    for (var threat_id in threats) {
        var threat = threats[threat_id];
        bad_directions.push(directionTo(my_ball, threat));
        var distance = getDistanceBetweenBalls(threat, my_ball);
        if ((ball.size / my_ball.size > 4 && distance < threat.size * 2.5) || distance < threat.size) {
            run_x.push(my_ball.x - (threat.x - my_ball.x));
            run_y.push(my_ball.y - (threat.y - my_ball.y));
            console.log("running from: " + threat + threat.size + ", direction: " + directionTo(my_ball, threat));
        }
    }

    /* Process virusses */
    for (var virus_id in virusses) {
        var virus = virusses[virus_id];
        if (getDistanceBetweenBalls(my_ball, virus) < my_ball.size) {
            run_x.push(my_ball.x - (virus.x - my_ball.x));
            run_y.push(my_ball.y - (virus.y - my_ball.y));
            console.log("running from: " + virus);
        }
    }

    var potential_target = null;
    var distance_target = 0;

    /* Process potential targets */
    for (var target_id in targets) {
        var target = targets[target_id];
        distance = getDistanceBetweenBalls(my_ball, target);
        if ((distance < my_ball.size * 3 && potential_target == null) || distance < distance_target) {
            potential_target = target;
            distance_target = distance;
        }
    }

    /* Do calculated action */
    if (run_x.length != 0) { // Running from something TODO corner and edge detection
        /* Calculate the average point to run away to and go there */
        client.moveTo(average(run_x), average(run_y));
    } else if (potential_target) {
        client.moveTo(potential_target.x, potential_target.y);
        var split_threshold = my_ball.size * 4;
        relativeSize = potential_target.size / my_ball.size;
        console.log("targetting: " + potential_target.name
                + ", distance: " + distance_target
                + ", threshold: " + split_threshold
                + ", relative size: " + relativeSize
                + ", balls left: " + client.my_balls.length);
        if (distance_target < split_threshold && 0.1 > relativeSize < 0.45 && client.my_balls.length < 2) { // TODO no biggies in the hood
            client.split();
        }
    } else {
        // Calculate right direction
        var willpower = calculateDirection(my_ball.x, my_ball.y, 10000, 10000);

        // TODO Favor going to the center of the map; edges are scary!
        // TODO Find pockets of food

        /* Process foods */
        potential_target = null;
        for (var food_id in foods) {
            var food = foods[food_id];
            distance = getDistanceBetweenBalls(my_ball, food);
            if (safeDirection(bad_directions, directionTo(my_ball, food)) && (potential_target == null || distance < distance_target)) {
                potential_target = food;
                distance_target = distance;
            }
        }
        if (potential_target != null) {
            //if (canvas) {
            //    //canvas.send(potential_target.x + ", " + potential_target.y);
            //    canvas.send(JSON.stringify([potential_target.x, potential_target.y]));
            //}
            client.moveTo(potential_target.x, potential_target.y);
        } else {
            client.moveTo(10000, 10000);
        }
    }
    //console.timeEnd('update');
}

/* #################### HELPER METHODS ####################### */

function getDistanceBetweenBalls(ball_1, ball_2) { //this calculates distance between 2 balls
    return Math.sqrt(Math.pow(ball_1.x - ball_2.x, 2) + Math.pow(ball_2.y - ball_1.y, 2));
}

function directionTo(ball1, ball2) {
    return calculateDirection(ball1.x, ball1.y, ball2.x, ball2.y);
}

function calculateDirection(x1, y1, x2, y2) {
    var theta = Math.atan2(y2 - y1, x2 - x1);
    if (theta < 0) theta += 2 * Math.PI;
    return theta * 180 / Math.PI;
}

/**
 * This function checks if the given direction is a safe direction to go to. This is done by checking if the direction
 * is too close near a BAD DIRECTION. In this case the method instantly returns false.The detection range is 160
 * degrees when there is one unsafe direction and 90 degrees otherwise..
 * @param list  List of unsafe directions
 * @param direction direction that will be checked
 * @returns {boolean}   true if safe
 */
function safeDirection(list, direction) {
    var range = list.length == 1 ? 80 : 45;
    for (var index in list) {
        var degrees = list[index];
        if (Math.abs(degrees - direction) < range || Math.abs(degrees - direction + 360) < range) return false;
    }
    return true;
}

/**
 * Calulates the average of all the items in a list. Used to calculate the average of multiple points, when the bot
 * wants to flee from threats.
 * @param list  The items
 * @returns {number}    Average of the items
 */
function average(list) {
    var total = 0;
    for (var i = 0; i < list.length; i++) {
        total += list[i];
    }
    return total / list.length;
}

/* ################### EVENT FUNCTIONS ################# */

client.on('mineBallDestroy', function (ball_id, reason) { //when my ball destroyed
    if (reason.by) {
        console.log(client.balls[reason.by] + ' ate my ball');
    }
    console.log('i lost my ball ' + ball_id + ', ' + client.my_balls.length + ' balls left');
    if(client.my_balls.length == 0) {
        setTimeout(function () {
            client.spawn("[xFake]Fieldhof");
        }, 2000);
    }
});


    /*
client.on('lostMyBalls', function () { //when i lost all my balls
    client.log('destroy spawning');
    setTimeout(function () {
        client.spawn("[xFake]");
    }, 2000);
});
*/



agario-devtools/player.js
var log_file = './records/demo.rec'; //record to play
var port = 411;	//port to listen for viewers

var fs = require('fs');
var WebSocket = require('ws');
var WebSocketServer = require('ws').Server;
var wss = new WebSocketServer({port: port});

wss.on('connection', function(wsc) {
    var viewer = new Viewer(wsc);

    wsc.on('close', function() {
        viewer.disconnected();
    });

    wsc.on('error', function() {
        viewer.disconnected();
    });
});

console.log('agar.io player started');
console.log('Open in browser http://agar.io/ and execute in console:');
console.log('connect("ws://127.0.0.1:' + port + '/");');
console.log('');
console.log('Waiting for connections...');

function Viewer(wsc) {
    this.abort = false;
    this.aborted = false;
    this.stream = fs.createReadStream(log_file);
    this.timer = 0;
    this.wsc = wsc;
    this.id = wsc.upgradeReq.connection.remoteAddress + ':' + wsc.upgradeReq.connection.remotePort;

    var that = this;

    this.log(this.id + ' connected');

    this.stream.once('readable', function() {
        if(that.isAborted()) return;

        that.processLine();
    });

}

Viewer.prototype = {
    isAborted: function() {
        if(this.aborted) return true;
        if(!this.abort) return false;

        this.stream.close();
        clearTimeout(this.timer);
        this.log(this.id + ' aborted');

        this.aborted = true;
        return true;
    },

    disconnected: function() {
        this.abort = true;
        this.isAborted();
    },

    processLine: function() {
        if(this.isAborted()) return;

        var header = this.stream.read(6);
        if(header === null) return this.log('end of tape');
        var time = header.readUInt16LE(0);
        var len = header.readUInt32LE(2);
        var packet_buffer = this.stream.read(len);


        if(this.wsc.readyState === WebSocket.OPEN) this.wsc.send(packet_buffer);
        this.timer = setTimeout(this.processLine.bind(this), time);
    },

    log: function(msg) {
        var now = new Date();
        var time = ('0' + now.getHours()).substr(-2) + ":"
            + ('0' + now.getMinutes()).substr(-2) + ":"
            + ('0' + now.getSeconds()).substr(-2);

        console.log(time + ' ' + this.id + ') ' + msg);
    }
};

agario-client/agario-client.js
var WebSocket = require('ws');
var Packet = require('./packet.js');
var EventEmitter = require('events').EventEmitter;

function Client(client_name) {
    //name used for log
    this.client_name = client_name;
    this.tick_counter = 0;

    if(this.debug >= 1)
        this.log('client created');
}

Client.prototype = {
    //you can change this values
    debug: 1, //debug level, 0-5 (5 will output extremely lot of data)
    inactive_destroy: 5*60*1000, //time in ms when to destroy inactive balls
    inactive_check: 10*1000, //time in ms when to search inactive balls

    //don't change things below if you don't understand what you're doing

    inactive_interval: 0, //ID of setInterval()
    balls: {},  //all balls
    my_balls: [], //IDs of my balls
    score: 0, //my score
    leaders: [], //IDs of leaders in FFA mode
    teams_scores: [], //scores of teams in Teams mode

    connect: function(server) {
        var headers = {
            'Origin': 'http://agar.io'
        };

        this.ws = new WebSocket(server, null, {headers: headers});
        this.ws.binaryType = "arraybuffer";
        this.ws.onopen = this.onConnect.bind(this);
        this.ws.onmessage = this.onMessage.bind(this);
        this.ws.onclose = this.onDisconnect.bind(this);
        this.ws.onerror = this.onError.bind(this);
        this.server = server;

        if(this.debug >= 1)
            this.log('connecting...');

        this.emit('connecting');
    },

    disconnect: function() {
        if(this.debug >= 1)
            this.log('disconnect() called');

        this.ws.close();
    },

    onConnect: function() {
        if(this.debug >= 1)
            this.log('connected');

        this.inactive_interval = setInterval(this.detsroyInactive.bind(this), this.inactive_check);

        var buf = new Buffer([255, 0, 0, 0, 0]);
        this.send(buf);

        this.emit('connected');
    },

    onError: function(e) {
        if(this.debug >= 1)
            this.log('connection error: ' + e);

        this.emit('connectionError', e);
        this.reset();
    },

    onDisconnect: function() {
        if(this.debug >= 1)
            this.log('disconnected');

        this.emit('disconnect');
        this.reset();
    },

    onMessage: function(e) {
        var packet = new Packet(e);
        var packet_id = packet.readUInt8();
        var processor = this.processors[packet_id];
        if(!processor) return this.log('warning: unknown packet ID(' + packet_id + '): ' + packet.toString());

        if(this.debug >= 4)
            this.log('ACK packet ID=' + packet_id + ' LEN=' + packet.length);
        if(this.debug >= 5)
            this.log('dump: ' + packet.toString());

        this.emit('message', packet);
        processor(this, packet);
    },

    send: function(buf) {
        if(this.debug >= 4)
            this.log('SEND packet ID=' + buf.readUInt8(0) + ' LEN=' + buf.length);

        if(this.debug >= 5)
            this.log('dump: ' + (new Packet(buf).toString()));

        this.ws.send(buf);
    },

    reset: function() {
        if(this.debug >= 3)
            this.log('reset()');

        this.leaders = [];
        this.teams_scores = [];
        this.my_balls = [];
        clearInterval(this.inactive_interval);
        for(var k in this.balls) if(this.balls.hasOwnProperty(k)) this.balls[k].destroy({'reason':'reset'});
        this.emit('reset');
    },

    detsroyInactive: function() {
        var time = (+new Date);

        if(this.debug >= 3)
            this.log('Destroying inactive balls');

        for(var k in this.balls) {
            if(!this.balls.hasOwnProperty(k)) continue;
            var ball = this.balls[k];
            if(time - ball.last_update < this.inactive_destroy) continue;
            if(ball.visible) continue;

            if(this.debug >= 3)
                this.log('Destroying inactive ' + ball);

            ball.destroy({reason: 'inactive'});
        }
    },

    processors: {
        //tick
        '16': function(client, packet) {
            var eaters_count = packet.readUInt16LE();

            client.tick_counter++;

            //reading eat events
            for(var i=0;i<eaters_count;i++) {
                var eater_id = packet.readUInt32LE();
                var eaten_id = packet.readUInt32LE();

                if(client.debug >= 4)
                    client.log(eater_id + ' ate ' + eaten_id + ' (' + client.balls[eater_id] + '>' + client.balls[eaten_id] + ')');

                if(!client.balls[eater_id]) new Ball(client, eater_id);
                client.balls[eater_id].update();
                if(client.balls[eaten_id]) client.balls[eaten_id].destroy({'reason':'eaten', 'by':eater_id});

                client.emit('somebodyAteSomething', eater_id, eaten_id);
            }


            //reading actions of balls
            while(1) {
                var is_virus = false;
                var ball_id;
                var coordinate_x;
                var coordinate_y;
                var size;
                var color;
                var nick = null;

                ball_id = packet.readUInt32LE();
                if(ball_id == 0) break;
                coordinate_x = packet.readSInt16LE();
                coordinate_y = packet.readSInt16LE();
                size = packet.readSInt16LE();

                var color_R = packet.readUInt8();
                var color_G = packet.readUInt8();
                var color_B = packet.readUInt8();

                color = (color_R << 16 | color_G << 8 | color_B).toString(16);
                color = '#' + ('000000' + color).substr(-6);

                var opt = packet.readUInt8();
                is_virus = !!(opt & 1);

                //reserved for future use?
                if (opt & 2) {
                    packet.offset += 4;
                }
                if (opt & 4) {
                    packet.offset += 8;
                }
                if (opt & 8) {
                    packet.offset += 16;
                }

                while(1) {
                    var char = packet.readUInt16LE();
                    if(char == 0) break;
                    if(!nick) nick = '';
                    nick += String.fromCharCode(char);
                }

                var ball = client.balls[ball_id] || new Ball(client, ball_id);
                ball.color = color;
                ball.virus = is_virus;
                ball.setCords(coordinate_x, coordinate_y);
                ball.setSize(size);
                if(nick) ball.setName(nick);
                ball.update_tick = client.tick_counter;
                ball.appear();
                ball.update();

                if(client.debug >= 4)
                    client.log('action: ball_id=' + ball_id + ' coordinate_x=' + coordinate_x + ' coordinate_y=' + coordinate_y + ' size=' + size + ' is_virus=' + is_virus + ' nick=' + nick);

                client.emit('ballAction', ball_id, coordinate_x, coordinate_y, size, is_virus, nick);
            }

            var balls_on_screen_count = packet.readUInt32LE();

            //disappear events
            for(i=0;i<balls_on_screen_count;i++) {
                ball_id = packet.readUInt32LE();

                ball = client.balls[ball_id] || new Ball(client, ball_id);
                ball.update_tick = client.tick_counter;
                ball.update();
                ball.disappear();
            }
        },

        '20': function() {
            //i dont know what is this
            //in original code it clears our balls array, but i never saw this packet
        },

        //new ID of your ball (when you join or press space)
        '32': function(client, packet) {
            var ball_id = packet.readUInt32LE();
            var ball = client.balls[ball_id] || new Ball(client, ball_id);
            ball.mine = true;
            if(!client.my_balls.length) client.score = 0;
            client.my_balls.push(ball_id);

            if(client.debug >= 2)
                client.log('my new ball: ' + ball_id);

            client.emit('myNewBall', ball_id);
        },

        //leaderboard update in FFA mode
        '49': function(client, packet) {
            var users = [];

            var count = packet.readUInt32LE();

            for(var i=0;i<count;i++) {
                var id = packet.readUInt32LE();

                var name = '';
                while(1) {
                    var char = packet.readUInt16LE();
                    if(char == 0) break;
                    name += String.fromCharCode(char);
                }

                users.push(id);
                var ball = client.balls[id] || new Ball(client, id);
                if(name) ball.setName(name);
                ball.update();
            }

            if(JSON.stringify(client.leaders) == JSON.stringify(users)) return;
            var old_leaders = client.leaders;
            client.leaders = users;

            if(client.debug >= 2)
                client.log('leaders update: ' + JSON.stringify(users));

            client.emit('leaderBoardUpdate', old_leaders, users);
        },

        //teams scored update in teams mode
        '50': function(client, packet) {
            var teams_count = packet.readUInt32LE();
            var teams_scores = [];

            for (var i=0;i<teams_count;++i) {
                teams_scores.push(packet.readFloat32LE());
            }

            if(JSON.stringify(client.teams_scores) == JSON.stringify(teams_scores)) return;
            var old_scores = client.teams_scores;

            if(client.debug >= 2)
                client.log('teams scores update: ' + JSON.stringify(teams_scores));

            client.teams_scores = teams_scores;

            client.emit('teamsScoresUpdate', old_scores, teams_scores);
        },

        //map size load
        '64': function(client, packet) {
            var min_x = packet.readFloat64LE();
            var min_y = packet.readFloat64LE();
            var max_x = packet.readFloat64LE();
            var max_y = packet.readFloat64LE();

            if(client.debug >= 2)
                client.log('map size: ' + [min_x, min_y, max_x, max_y].join(','));

            client.emit('mapSizeLoad', min_x, min_y, max_x, max_y);
        },

        //another unknown backet
        '72': function() {
            //packet is sent by server but not used in original code
        },

        '240': function(client, packet) {
            packet.offset += 4;
            var packet_id = packet.readUInt8();
            var processor = client.processors[packet_id];
            if(!processor) return client.log('warning: unknown packet ID(240->' + packet_id + '): ' + packet.toString());
            processor(client, packet);
        },

        //somebody won, end of the game (server restart)
        '254': function(client) {
            if(client.debug >= 1)
                client.log(client.balls[client.leaders[0]] + ' WON THE GAME! Server going for restart');

            client.emit('winner', client.leaders[0]);
        }
    },

    updateScore: function() {
        var potential_score = 0;
        for (var i=0;i<this.my_balls.length;i++) {
            var ball_id = this.my_balls[i];
            var ball = this.balls[ball_id];
            potential_score += Math.pow(ball.size, 2);
        }
        var old_score = this.score;
        var new_score = Math.max(this.score, Math.floor(potential_score / 100));

        if (this.score == new_score) return;
        this.score = new_score;
        this.emit('scoreUpdate', old_score, new_score);

        if(this.debug >= 2)
            console.log('score: ' + new_score);

    },

    log: function(msg) {
        console.log(this.client_name + ': ' + msg);
    },

    //functions that you can call to control your balls

    //spawn ball
    spawn: function(name) {
        var buf = new Buffer(1 + 2*name.length);
        buf.writeUInt8(0, 0);
        for (var i=0;i<name.length;i++) {
            buf.writeUInt16LE(name.charCodeAt(i), 1 + i*2);
        }
        this.send(buf);
    },

    moveTo: function(x, y) {
        var buf = new Buffer(21);
        buf.writeUInt8(16, 0);
        buf.writeDoubleLE(x, 1);
        buf.writeDoubleLE(y, 9);
        buf.writeUInt32LE(0, 17);
        this.send(buf);
    },

    //split your balls
    //they will split in direction that you have set with moveTo()
    split: function() {
        var buf = new Buffer([17]);
        this.send(buf);
    },

    //eject some mass
    //mass will eject in direction that you have set with moveTo()
    eject: function() {
        var buf = new Buffer([21]);
        this.send(buf);
    }
};

function Ball(client, id) {
    if(client.balls[id]) return client.balls[id];

    this.id = id;
    this.name = null;
    this.x = 0;
    this.y = 0;
    this.size = 0;
    this.virus = false;
    this.mine = false;

    this.client = client;
    this.destroyed = false;
    this.visible = false;
    this.last_update = (+new Date);
    this.update_tick = 0;

    client.balls[id] = this;
    return this;
}
Ball.prototype = {
    destroy: function(reason) {
        this.destroyed = reason;
        delete this.client.balls[this.id];
        var mine_ball_index = this.client.my_balls.indexOf(this.id);
        if(mine_ball_index > -1) {
            this.client.my_balls.splice(mine_ball_index, 1);
            this.client.emit('mineBallDestroy', this.id, reason);
            if(!this.client.my_balls.length) this.client.emit('lostMyBalls');
        }

        this.emit('destroy', reason);
        this.client.emit('ballDestroy', this.id, reason);
    },

    setCords: function(new_x, new_y) {
        if(this.x == new_x && this.y == new_y) return;
        var old_x = this.x;
        var old_y = this.y;
        this.x = new_x;
        this.y = new_y;

        if(!old_x && !old_y) return;
        this.emit('move', old_x, old_y, new_x, new_y);
        this.client.emit('ballMove', this.id, old_x, old_y, new_x, new_y);
    },

    setSize: function(new_size) {
        if(this.size == new_size) return;
        var old_size = this.size;
        this.size = new_size;

        if(!old_size) return;
        this.emit('resize', old_size, new_size);
        this.client.emit('ballResize', this.id, old_size, new_size);
        if(this.mine) this.client.updateScore();
    },

    setName: function(name) {
        if(this.name == name) return;
        var old_name = this.name;
        this.name = name;

        this.emit('rename', old_name, name);
        this.client.emit('ballRename', this.id, old_name, name);
    },

    update: function() {
        var old_time = this.last_update;
        this.last_update = (+new Date);

        this.emit('update', old_time, this.last_update);
        this.client.emit('ballUpdate', this.id, old_time, this.last_update);
    },

    appear: function() {
        if(this.visible) return;
        this.visible = true;
        this.emit('appear');
        this.client.emit('ballAppear', this.id);

        if(this.mine) this.client.updateScore();
    },

    disappear: function() {
        if(!this.visible) return;
        this.visible = false;
        this.emit('disappear');
        this.client.emit('ballDisppear', this.id);
    },

    toString: function() {
        if(this.name) return this.id + '(' + this.name + ')';
        return this.id.toString();
    }
};

// Inherit from EventEmitter
for (var key in EventEmitter.prototype) {
    if(!EventEmitter.prototype.hasOwnProperty(key)) continue;
    Client.prototype[key] = Ball.prototype[key] = EventEmitter.prototype[key];
}

//deprecated methods on 04.06.2015
Client.prototype.off = Ball.prototype.off = function() {
    if(!this._off_notified)  {
        console.trace('agario-client: .off() is deprecated, use ".removeListener()" instead.\n' +
            'Please change your code, here is stack trace for you');
        this._off_notified=true;
    }
    this.removeListener.apply(this, arguments);
};
Client.prototype.offAll = Ball.prototype.offAll = function() {
    if(!this._offAll_notified)  {
        console.trace('agario-client: .offAll() is deprecated, use ".removeAllListeners()" instead.\n' +
            'Please change your code, here is stack trace for you');
        this._offAll_notified=true;
    }
    this.removeAllListeners.apply(this, arguments);
};
Client.prototype.emitEvent = Ball.prototype.emitEvent = function() {
    if(!this._emitEvent_notified)  {
        console.trace('agario-client: .emitEvent() is deprecated, use ".emit()" instead.\n' +
            'Please change your code, here is stack trace for you');
        this._emitEvent_notified=true;
    }
    this.emit.apply(this, arguments);
};

module.exports = Client;
